### PR TITLE
Add square conformal projections from libproject

### PR DESCRIPTION
**Pull request by kbevers**
This PR adds five new projections to PROJ:

`adams_hemi`: Adams Hemisphere in a Square
`adams_ws1`:  Adams World in a Square I
`adams_ws2`: Adams World in a Square II
`guyou`:      Guyou
`peirce_q`:   Pierce Quincuncial

The code originates from Gerry Evendens libproject and has been adapted
to work with modern PROJ. To ensure that the modified code works as
intended extensive test data has been created using libproject and
sproj to ensure that no errors occured when porting from libproject to PROJ.
The test data is wrapped in a gie files. All test cases reproduce
results from libproject at the mm level.

Docs are still missing at this point. Will be added before merging.

This work is a first step towards implementing the Spilhaus projection as requested in #1851. 

 - [ ] Closes #xxxx
 - [x] Tests added
 - [x] Added clear title that can be used to generate release notes
 - [x] Fully documented, including updating `docs/source/*.rst` for new API

### CODE REVIEW

File: src/projections/adams.cpp [hunk 1]
```diff
@@ -0,0 +1,217 @@
+/*
+* Implementation of the Guyou, Pierce Quincuncial, Adams Hemisphere in a Square,
+* Adams World in a Square I & II projections.
+*
+* Based on original code from libproj4 written by Geral Evenden. Adapted by to
```
Comment (rouault) [inline src/projections/adams.cpp]:
```suggestion
* Based on original code from libproj4 written by Geral Evenden. Adapted to
```

File: src/projections/adams.cpp [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+*
+* Copyright (c) 2005, 2006, 2009   Gerald I. Evenden
+* Copyright (c) 2020 Kristian Evers
+*
+* Related material
+* ----------------
+*
+*  CONFORMAL PROJECTION OF THE SPHERE WITHIN A SQUARE, 1929, OSCAR S. ADAMS,
+*  U.S.COASTAND GEODETIC SURVEY, Special Publication No.153,
+*  ftp://ftp.library.noaa.gov/docs.lib/htdocs/rescue/cgs_speint cpubs/QB275U35no1531929.pdf
```
Comment (rouault) [inline src/projections/adams.cpp]:
```suggestion
*  ftp://ftp.library.noaa.gov/docs.lib/htdocs/rescue/cgs_specpubs/QB275U35no1531929.pdf
```

File: src/projections/adams.cpp [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
+            a = cos(aasin(P->ctx, spp)) * sin(0.5 * lp.lam);
+            sm = (spp + a) < 0.;
+            sn = (spp - a) < 0.;
+            b = aacos(P->ctx, spp);
+            a = aacos(P->ctx, a);
+        }
+        break;
+    }
+
+    m = aasin(P->ctx, sqrt(fabs(1. + cos(a + b))));
```
Comment (rouault) [inline src/projections/adams.cpp]: the fabs() seems unnecessary here. Unless due to rare floating point issue 1+cos() would be negative ?

File: src/projections/adams.cpp [hunk 4]
```diff
@@ -0,0 +1,217 @@
...
+        double temp = d1;
+        d1 = y2 * d1 - d2 + *Cp--;
+        d2 = temp;
+    }
+
+    return phi * (y * d1 - d2 + 0.5 * *Cp);
+}
+
+static PJ_XY forward(PJ_LP lp, PJ *P) {
+    double m, n, a=0., b=0.;
```
Comment (rouault) [inline src/projections/adams.cpp]: m and n could be declared much later, where they are actually set

File: src/projections/adams.cpp [hunk 5]
```diff
@@ -0,0 +1,217 @@
...
+    int sm=0, sn=0;
```
Comment (rouault) [inline src/projections/adams.cpp]:
```suggestion
    bool sm=false, sn=false;
```

File: src/projections/adams.cpp (hunk 3 above)
Comment (kbevers) [reply src/projections/adams.cpp]: At least it seems to be unnecessary. I don't know about floating point weirdness. Once I've got successful CI builds I'll try to remove fabs() here and see if it makes a difference.

File: src/projections/adams.cpp (hunk 2 above)
Comment (kbevers) [reply src/projections/adams.cpp]: Do, nice catch!

File: src/projections/adams.cpp (hunk 3 above)
Comment (kbevers) [reply src/projections/adams.cpp]: Removing fabs works locally at least

File: src/projections/adams.cpp [hunk 6]
```diff
@@ -0,0 +1,217 @@
...
+static PJ_XY forward(PJ_LP lp, PJ *P) {
+    double a=0., b=0.;
+    bool sm=false, sn=false;
+    PJ_XY xy;
+    struct pj_opaque *Q = static_cast<struct pj_opaque*>(P->opaque);
+
+    switch (Q->mode) {
+    case GUYOU:
+        if ((fabs(lp.lam) - TOL) > M_PI_2) {
+            proj_errno_set(P, PJD_ERR_TOLERANCE_CONDITION);
```
Comment (rouault) [inline src/projections/adams.cpp:100]: no return ?

File: src/projections/adams.cpp [hunk 7]
```diff
@@ -0,0 +1,217 @@
...
+        }
+
+        if (fabs(fabs(lp.phi) - M_PI_2) < TOL) {
+            xy.x = 0;
```
Comment (rouault) [inline src/projections/adams.cpp:105]: Case not tested according to https://coveralls.io/builds/30051762/source?filename=src/projections/adams.cpp

File: src/projections/adams.cpp [hunk 8]
```diff
@@ -0,0 +1,217 @@
...
+            a = aacos(P->ctx, cp * (sl + cl) * RSQRT2);
+            b = aacos(P->ctx, cp * (sl - cl) * RSQRT2);
+            sm = sl < 0.;
+            sn = cl > 0.;
+        }
+        break;
+    case ADAMS_HEMI: {
+            double sp = sin(lp.phi);
+            if ((fabs(lp.lam) - TOL) > M_PI_2) {
+                proj_errno_set(P, PJD_ERR_TOLERANCE_CONDITION);
```
Comment (rouault) [inline src/projections/adams.cpp:131]: no return ?

File: src/projections/adams.cpp (hunk 6 above)
Comment (kbevers) [reply src/projections/adams.cpp:100]: Wasn't in the original code. I guess it should return `proj_coord_error().xy`. I assume that the code after this check results in at least one `HUGE_VAL`s being returned but it is better to bail out right away. Same goes for the other comment.

Comment (kbevers) [reply src/projections/adams.cpp:100]: Oh wait, `F_ERROR` magic probably made sure to return something sensible here.

**Comment by mwtoews** [comment]
I see that `adams_wsI` and `adams_wsII` are the names that Gerry originally had ([look here](https://github.com/rouault/libproj4/search?q=adams&unscoped_q=adams)), however this naming scheme with upper case Roman numerals is different than any other projection name, which are all lowercase with Arabic numerals (e.g. `wink1` for "Winkel I", or find them all with `git grep "PROJ_HEAD("`).

Should these projection names be `adams_ws1` and `adams_ws2` to follow similar projection names in use?

**Comment by kbevers** [comment]
> Should these projection names be adams_ws1 and adams_ws2 to follow similar projection names in use?

That's a good question. Consistency is probably better than libproject compatibility (don't think it was ever used much).

**Comment by nyalldawson** [comment]
This is great stuff, nice work @kbevers!

**Comment by rouault** [comment]
Looking at https://raw.githubusercontent.com/Esri/projection-engine-db-doc/master/csv/pe_list_projcs.csv, one can see the following:
```
54098,54098,"PE_PCS_WORLD_ADAMS_SQ_2","WGS_1984_Adams_Square_II","PROJCS[""WGS_1984_Adams_Square_II"",GEOGCS[""GCS_WGS_1984"",DATUM[""D_WGS_1984"",SPHEROID[""WGS_1984"",6378137.0,298.257223563]],PRIMEM[""Greenwich"",0.0],UNIT[""Degree"",0.0174532925199433]],PROJECTION[""Adams_Square_II""],PARAMETER[""False_Easting"",0.0],PARAMETER[""False_Northing"",0.0],PARAMETER[""Scale_Factor"",1.0],PARAMETER[""Azimuth"",0.0],PARAMETER[""Longitude_Of_Center"",0.0],PARAMETER[""Latitude_Of_Center"",0.0],PARAMETER[""XY_Plane_Rotation"",0.0],UNIT[""Meter"",1.0]]","Adams Square II","Esri","10.8.0","no","World",-90.0,90.0,-180.0,180.0
54099,54099,"PE_PCS_WORLD_SPILHAUS_OCEAN_MAP","WGS_1984_Spilhaus_Ocean_Map_in_Square","PROJCS[""WGS_1984_Spilhaus_Ocean_Map_in_Square"",GEOGCS[""GCS_WGS_1984"",DATUM[""D_WGS_1984"",SPHEROID[""WGS_1984"",6378137.0,298.257223563]],PRIMEM[""Greenwich"",0.0],UNIT[""Degree"",0.0174532925199433]],PROJECTION[""Adams_Square_II""],PARAMETER[""False_Easting"",0.0],PARAMETER[""False_Northing"",0.0],PARAMETER[""Scale_Factor"",1.0],PARAMETER[""Azimuth"",40.17823482],PARAMETER[""Longitude_Of_Center"",66.94970198],PARAMETER[""Latitude_Of_Center"",-49.56371678],PARAMETER[""XY_Plane_Rotation"",45.0],UNIT[""Meter"",1.0]]","Spilhaus World Ocean Map in a Square","Esri","10.8.0","no","World",-90.0,90.0,-180.0,180.0
```
So Esri has added an explicit XY_Plane_Rotation parameter, that they set to 0 for the WGS_1984_Adams_Square_II, and to 45 for Spilhaus.
libproject hardcodes 45 for . Should we have a rot_xy parameter (like in healpix ?). That parameter actually could potentially be handled in a generic way for all projections.

File: src/pj_list.h [hunk 9]
```diff
@@ -115,6 +119,7 @@ PROJ_HEAD(ortel, "Ortelius Oval")
 PROJ_HEAD(ortho, "Orthographic")
 PROJ_HEAD(pconic, "Perspective Conic")
 PROJ_HEAD(patterson, "Patterson Cylindrical")
+PROJ_HEAD(peirce_q, "Pierce Quincuncial")
```
Comment (busstoptaktik) [inline src/pj_list.h]: Pierce -> Peirce.

Charles Sanders Peirce: American semioticist, logician, geodesist and much more. Father of pragmaticism and abductive reasoning. Overall interesting figure (look him up one day you have too much time on your hands)

File: src/projections/adams.cpp [hunk 10]
```diff
@@ -0,0 +1,219 @@
+/*
+* Implementation of the Guyou, Pierce Quincuncial, Adams Hemisphere in a Square,
+* Adams World in a Square I & II projections.
+*
+* Based on original code from libproj4 written by Geral Evenden. Adapted to modern
```
Comment (busstoptaktik) [inline src/projections/adams.cpp]: Geral -> Gerald

File: src/projections/adams.cpp [hunk 11]
```diff
@@ -0,0 +1,219 @@
...
+* for reference.
+*
+* Copyright (c) 2005, 2006, 2009   Gerald I. Evenden
+* Copyright (c) 2020 Kristian Evers
+*
+* Related material
+* ----------------
+*
+*  CONFORMAL PROJECTION OF THE SPHERE WITHIN A SQUARE, 1929, OSCAR S. ADAMS,
+*  U.S.COASTAND GEODETIC SURVEY, Special Publication No.153,
```
Comment (busstoptaktik) [inline src/projections/adams.cpp]: COASTAND -> COAST AND

**Review by busstoptaktik — COMMENTED** [review]
I do not have much to contribute, but a few "speeling missteaks" is probably also worth correcting...

**Comment by kbevers** [comment]
> So Esri has added an explicit XY_Plane_Rotation parameter, that they set to 0 for the WGS_1984_Adams_Square_II, and to 45 for Spilhaus.

The Adams square II is actually subject to a 45 rotation as seen here:
https://github.com/OSGeo/PROJ/blob/472712edfccafc62ddeb67d4a5e67fd09666c7fb/src/projections/adams.cpp#L174-L178

So for the Spilhaus perhaps this should just be a 90 degree rotation? Or 0? Not sure of the direction here...

But yeah, a generic rotation parameter would also work. Are you aware of other cases where a parameter like that is used by Esri or others?

**Comment by rouault** [comment]
> So for the Spilhaus perhaps this should just be a 90 degree rotation? Or 0? Not sure of the direction here...

Hum, I'm not sure. A possibility would be that xy_rot=0 would skip the current rotation (if that's what Esri WGS_1984_Adams_Square_II does ?), and xy_rot=45 should apply it (Spilhaus use case ?). Would be good if someone could check against a Esri install (I don't have one) what they get for the above 2 definitions (and possibly modify the base WGS_1984_Adams_Square_II one, to just test different values of the rotation parameter), so we are consistent with what they do.

> Are you aware of other cases where a parameter like that is used by Esri or others?

Digging into Esri public repo of definitions, I see that they use it at least for Krovak (where it can be 0 or 90, as long as X_Scale and Y_Scale to apply axis reversal) and Oblique Mercator (where it maps to PROJ +gamma parameter)

**Comment by nyalldawson** [comment]
> Would be good if someone could check against a Esri install 

Can you clarify exactly what you're after? I can test tomorrow

**Comment by rouault** [comment]
> Can you clarify exactly what you're after? I can test tomorrow

For example, take a world map (Natural Earth or whatever) and reproject it to ESRI:54098 "WGS_1984_Adams_Square_II" and ESRI:54099 "WGS_1984_Spilhaus_Ocean_Map_in_Square"

And if that's possible test reprojecting to the following custom variations
```
PROJCS["WGS_1984_Adams_Square_II_with_XY_rotation_20",GEOGCS["GCS_WGS_1984",DATUM["D_WGS_1984",SPHEROID["WGS_1984",6378137.0,298.257223563]],PRIMEM["Greenwich",0.0],UNIT["Degree",0.0174532925199433]],PROJECTION["Adams_Square_II"],PARAMETER["False_Easting",0.0],PARAMETER["False_Northing",0.0],PARAMETER["Scale_Factor",1.0],PARAMETER["Azimuth",0.0],PARAMETER["Longitude_Of_Center",0.0],PARAMETER["Latitude_Of_Center",0.0],PARAMETER["XY_Plane_Rotation",20.0]

PROJCS["WGS_1984_Adams_Square_II_with_XY_rotation_45",GEOGCS["GCS_WGS_1984",DATUM["D_WGS_1984",SPHEROID["WGS_1984",6378137.0,298.257223563]],PRIMEM["Greenwich",0.0],UNIT["Degree",0.0174532925199433]],PROJECTION["Adams_Square_II"],PARAMETER["False_Easting",0.0],PARAMETER["False_Northing",0.0],PARAMETER["Scale_Factor",1.0],PARAMETER["Azimuth",0.0],PARAMETER["Longitude_Of_Center",0.0],PARAMETER["Latitude_Of_Center",0.0],PARAMETER["XY_Plane_Rotation",45.0]
```

**Comment by kbevers** [comment]
I think this PR is more or less finished now. Thanks for all the input, it's been super helpful!

File: src/projections/adams.cpp [hunk 12]
```diff
@@ -0,0 +1,215 @@
...
+PROJ_HEAD(adams_ws2, "Adams World in a Square II") "\n\tMisc Sph No inv";
+
+namespace { // anonymous namespace
+
+enum projection_type {
+    GUYOU,
+    PEIRCE_Q,
+    ADAMS_HEMI,
+    ADAMS_WSI,
+    ADAMS_WSII,
```
Comment (mwtoews) [inline src/projections/adams.cpp]: Perhaps also make these `ADAMS_WS1` and `ADAMS_WS2` too. Here and below.

**Comment by nyalldawson** [comment]
@rouault 

> , take a world map (Natural Earth or whatever) and reproject it to ESRI:54098

image

> and ESRI:54099 "WGS_1984_Spilhaus_Ocean_Map_in_Square"

image

> ```PROJCS["WGS_1984_Adams_Square_II_with_XY_rotation_20",GEOGCS["GCS_WGS_1984",DATUM["D_WGS_1984",SPHEROID["WGS_1984",6378137.0,298.257223563]],PRIMEM["Greenwich",0.0],UNIT["Degree",0.0174532925199433]],PROJECTION["Adams_Square_II"],PARAMETER["False_Easting",0.0],PARAMETER["False_Northing",0.0],PARAMETER["Scale_Factor",1.0],PARAMETER["Azimuth",0.0],PARAMETER["Longitude_Of_Center",0.0],PARAMETER["Latitude_Of_Center",0.0],PARAMETER["XY_Plane_Rotation",20.0]```

It refuses to import either of these custom WKT strings

**Comment by rouault** [comment]
@nyalldawson OK, so comparing images, at least for  ESRI:54098 "WGS_1984_Adams_Square_II", with XY_Plane_Rotation=0, they do apply the 45° final rotation that PROJ also applies. Thus XY_Plane_Rotation must be an extra rotation on top of that. Interesting that it refuses to import custom WKT with other XY_Plane_Rotation values. Smells like only ESRI:54098 and ESRI:54099 cases have been somehow hardcoded.

**Comment by nyalldawson** [comment]
@rouault there's always a chance I was importing wrong. There's zero documentation for this and no error messages - just no result after the import. ESRI documentation is just plain woeful.

**Pull request merged**